rust/src/storage/mod.rs

```diff
@@ -1,600 +1,312 @@
-    #[inline]
-    pub fn path(&self) -> String {
-        match self {
-            Uri::LocalPath(x) => x.to_string(),
-            #[cfg(any(feature = "s3", feature = "s3-rustls"))]
-            Uri::S3Object(x) => x.key.to_string(),
-            #[cfg(feature = "azure")]
-            Uri::AdlsGen2Object(x) => x.path.to_string(),
-            #[cfg(feature = "gcs")]
-            Uri::GCSObject(x) => x.path.to_string(),
-        }
+    /// convert a table [Path] to a fully qualified uri
+    pub fn to_uri(&self, location: &Path) -> String {
+        let uri = match self.storage_url.scheme() {
+            "file" | "" => {
+                // On windows the drive (e.g. 'c:') is part of root and must not be prefixed.
+                #[cfg(windows)]
+                let os_uri = format!("{}/{}", self.storage_url.prefix, location.as_ref());
+                #[cfg(unix)]
+                let os_uri = format!("/{}/{}", self.storage_url.prefix, location.as_ref());
+                os_uri
+            }
+            _ => format!("{}/{}", self.storage_url.as_str(), location.as_ref()),
+        };
+        uri.trim_end_matches('/').into()
     }
 }
 
-/// Parses the URI and returns a variant of the Uri enum for the appropriate storage backend based
-/// on scheme.
-pub fn parse_uri<'a>(path: &'a str) -> Result<Uri<'a>, UriError> {
-    let parts: Vec<&'a str> = path.split("://").collect();
+/// Object Store implementation for DeltaTable.
+///
+/// The [DeltaObjectStore] implements the [object_store::ObjectStore] trait to facilitate
+/// interoperability with the larger rust / arrow ecosystem. Specifically it can directly
+/// be registered as store within datafusion.
+///
+/// The table root is treated as the root of the object store.
+/// All [Path] are reported relative to the table root.
+#[derive(Debug, Clone)]
+pub struct DeltaObjectStore {
+    storage: Arc<DynObjectStore>,
+    config: DeltaObjectStoreConfig,
+}
 
-    if parts.len() == 1 {
-        return Ok(Uri::LocalPath(parts[0]));
+impl std::fmt::Display for DeltaObjectStore {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "DeltaObjectStore({})", self.config.storage_url.as_str())
     }
+}
 
-    match parts[0] {
-        "s3" => {
-            cfg_if::cfg_if! {
-                if #[cfg(any(feature = "s3", feature = "s3-rustls"))] {
-                    let mut path_parts = parts[1].splitn(2, '/');
-                    let bucket = match path_parts.next() {
-                        Some(x) => x,
-                        None => {
-                            return Err(UriError::MissingObjectBucket);
-                        }
-                    };
-                    let key = match path_parts.next() {
-                        Some(x) => x,
-                        None => {
-                            return Err(UriError::MissingObjectKey);
-                        }
-                    };
+impl DeltaObjectStore {
```

Comment:
Exactly! Since all the paths from the log are relative, we always have to do that conversion, so I thought it would be easiest to just have a base path inside the object store. There are some convenience functions defined, but without the base path, these would certainly not warrant wrapping the store.